File: src/classes/Ticket.js
import { Status, Priority } from './Enums';
import User from './User';

class Ticket {
  constructor(ticketId, title, description, status, priority, isOpen = true) {
    this._ticketId = ticketId;
    this._title = title;
    this._description = description;
    this._status = status;
    this._priority = priority;
    this._isOpen = isOpen;
    this._createdAt = new Date();
    this._updatedAt = new Date();
    this._createdBy = null;
    this._assignedTo = null;
  }

  get ticketId() {
    return this._ticketId;
  }

  get title() {
    return this._title;
  }

  set title(newTitle) {
    this._title = newTitle;
    this._updateTimestamp();
  }

  get description() {
    return this._description;
  }

  set description(newDescription) {
    this._description = newDescription;
    this._updateTimestamp();
  }

  get status() {
    return this._status;
  }

  set status(newStatus) {
    this._status = newStatus;
    this._updateTimestamp();
  }

  get priority() {
    return this._priority;
  }

  set priority(newPriority) {
    this._priority = newPriority;
    this._updateTimestamp();
  }

  get isOpen() {
    return this._isOpen;
  }

  set isOpen(newIsOpen) {
    this._isOpen = newIsOpen;
    this._updateTimestamp();
  }

  get createdAt() {
    return this._createdAt;
  }

  get updatedAt() {
    return this._updatedAt;
  }

  get createdBy() {
    return this._createdBy;
  }

  set createdBy(user) {
    this._createdBy = user;
    this._updateTimestamp();
  }

  get assignedTo() {
    return this._assignedTo;
  }

  set assignedTo(user) {
    this._assignedTo = user;
    this._updateTimestamp();
  }

  updateStatus(newStatus) {
    this.status = newStatus;
    return this;
  }

  close() {
    this.isOpen = false;
    this.status = Status.Done;
    return this;
  }

  open() {
    this.isOpen = true;
    if (this.status === Status.Done) {
      this.status = Status.ToDo;
    }
    return this;
  }

  isOverdue() {
    if (!this.isOpen) return false;
    const now = new Date();
    const daysSinceCreation = Math.floor((now - this.createdAt) / (1000 * 60 * 60 * 24));
    return daysSinceCreation > 7;
  }

  getAgeInDays() {
    const now = new Date();
    return Math.floor((now - this.createdAt) / (1000 * 60 * 60 * 24));
  }

  _updateTimestamp() {
    this._updatedAt = new Date();
  }

  toDict() {
    return {
      ticket_id: this.ticketId,
      title: this.title,
      description: this.description,
      status: this.status,
      priority: this.priority,
      is_open: this.isOpen,
      created_at: this.createdAt.toISOString(),
      updated_at: this.updatedAt.toISOString(),
      created_by: this.createdBy ? this.createdBy.toDict() : null,
      assigned_to: this.assignedTo ? this.assignedTo.toDict() : null
    };
  }

  static fromDict(data) {
    const ticket = new Ticket(
      data.ticket_id,
      data.title,
      data.description,
      data.status,
      data.priority,
      data.is_open
    );

    if (data.created_at) {
      ticket._createdAt = new Date(data.created_at);
    }
    if (data.updated_at) {
      ticket._updatedAt = new Date(data.updated_at);
    }
    if (data.created_by) {
      ticket.createdBy = User.fromDict(data.created_by);
    }
    if (data.assigned_to) {
      ticket.assignedTo = User.fromDict(data.assigned_to);
    }

    return ticket;
  }
}

export default Ticket; 

File: TicketingSystem/Objects/ticket.py
from datetime import datetime
from typing import Optional

from .enums import Status, Priority
from .user import User

class Ticket:
  def __init__(
    self,
    ticketId: int,
    title: str,
    description: str,
    status: Status,
    priority: Priority,
    isOpen: bool = true
  ):
    
    self.ticket_id = ticket_id
    self.title = title
    self.description = description
    self.status = status
    self.priority = priority
    self.created_at = datetime.now()
    self.updated_at = datetime.now()
    self.is_open = is_open
    self.created_by: Optional[User] = None
    self.assigned_to: Optional[User] = None

    self._validate()

  def _validate(self):
    if not self.title:
      raise ValueError("Title cannot be empty")
    if not self.description:
      raise ValueError("Description cannot be empty")
    if not self.org_id:
      raise ValueError("Organization ID cannot be empty")

  def to_dict(self):
    return {
    "ticket_id": self.ticket_id,
    "title": self.title,
    "description": self.description,
    "status": self.status.value,
    "priority": self.priority.value,
    "created_at": self.created_at.isoformat(),
    "updated_at": self.updated_at.isoformat(),
    "is_open": self.is_open,
    "assigned_to": self.assigned_to
    }
    
